File: workers/worker-mysql/src/worker_mysql/registry.py
import os

from openregex_libs.models import (
    CheatSheetCategory,
    CheatSheetItem,
    EngineCapabilities,
    EngineDocs,
    EngineExample,
    EngineFlag,
    EngineInfo,
    WorkerInfo,
    WORKER_SCHEMA_VERSION,
)

from .engine import FLAG_METADATA

WORKER_NAME = "worker-mysql"
WORKER_VERSION = os.environ.get("WORKER_VERSION", "Unknown")
WORKER_RELEASE_DATE = os.environ.get("WORKER_RELEASE_DATE", "Unreleased")

CAT_CLASSES = "Character Classes & Escapes"
CAT_ANCHORS = "Anchors & Boundaries"
CAT_QUANTIFIERS = "Quantifiers"
CAT_GROUPS = "Grouping & Backreferences"
CAT_ADVANCED = "Lookarounds & Advanced"
CAT_SQL = "SQL Surface"


def build_engine_flags(*names: str) -> list[EngineFlag]:
    flags: list[EngineFlag] = []
    for name in names:
        description, group = FLAG_METADATA.get(name, (f"Flag ({name})", "Basic"))
        flags.append(EngineFlag(name=name, description=description, group=group))
    return flags


def build_engine(server_version: str) -> EngineInfo:
    major_minor = ".".join(server_version.split(".")[:2])
    return EngineInfo(
        engine_id="mysql_icu",
        engine_language_type="SQL",
        engine_language_version=server_version,
        engine_regex_lib="ICU",
        engine_regex_lib_version=f"bundled with MySQL {server_version}",
        engine_label=f"MySQL {major_minor} (ICU)",
        engine_capabilities=EngineCapabilities(
            flags=build_engine_flags("c", "i", "m", "n", "u"),
            supports_lookaround=True,
            supports_backrefs=True,
        ),
        engine_docs=EngineDocs(
            trivia=[
                "MySQL 8.0 threw out the old Spencer engine and switched to ICU, so patterns written for MySQL 5.7 can behave differently on 8.0.",
                "MySQL is distributed under GPL-2.0 with the FOSS License Exception; ICU ships under the Unicode License.",
                "ICU regex is the same engine family behind Swift's NSRegularExpression and Java's ICU4J, so results here are a good proxy for those.",
                "MariaDB is NOT a drop-in twin here: it kept PCRE2 instead of moving to ICU, so the two forks disagree on syntax and semantics.",
                "The SQL surface exposes match positions (REGEXP_INSTR) but no capture-group positions, so this engine reports whole matches only.",
                "regexp_time_limit caps ICU's internal step count. Its default of 32 is so tight that even moderate patterns fail; this worker raises it and adds a wall-clock SLA on top.",
                "Case sensitivity follows the collation, not the pattern: with the default utf8mb4_0900_ai_ci everything matches case-insensitively. This worker forces an accent- and case-sensitive collation so the flags mean what they say.",
            ],
            cheat_sheet_url="https://dev.mysql.com/doc/refman/8.4/en/regexp.html",
        ),
        engine_cheat_sheet=[
            CheatSheetCategory(
                category=CAT_CLASSES,
                items=[
                    CheatSheetItem(character=".", description="Any character except newline unless the 'n' flag is set"),
                    CheatSheetItem(character="\\w", description="Word character (SQL string literals need the backslash doubled)"),
                    CheatSheetItem(character="\\W", description="Non-word character"),
                    CheatSheetItem(character="\\d", description="Decimal digit"),
                    CheatSheetItem(character="\\D", description="Non-digit"),
                    CheatSheetItem(character="\\s", description="Whitespace character"),
                    CheatSheetItem(character="\\S", description="Non-whitespace character"),
                    CheatSheetItem(character="\\p{L}", description="Unicode property class"),
                    CheatSheetItem(character="\\X", description="Extended grapheme cluster"),
                    CheatSheetItem(character="[[:alpha:]]", description="POSIX character class"),
                    CheatSheetItem(character="[a-z]", description="Character range"),
                    CheatSheetItem(character="[^a-z]", description="Negated character class"),
                    CheatSheetItem(character="[\\p{L}&&[^\\p{Lu}]]", description="Character class intersection (ICU set syntax)"),
                ],
            ),
            CheatSheetCategory(
                category=CAT_ANCHORS,
                items=[
                    CheatSheetItem(character="^", description="Start of string, or start of line under the 'm' flag"),
                    CheatSheetItem(character="$", description="End of string, or end of line under the 'm' flag"),
                    CheatSheetItem(character="\\A", description="Start of input"),
                    CheatSheetItem(character="\\z", description="End of input"),
                    CheatSheetItem(character="\\Z", description="End of input, ignoring a final line terminator"),
                    CheatSheetItem(character="\\b", description="Word boundary"),
                    CheatSheetItem(character="\\B", description="Non-word boundary"),
                    CheatSheetItem(character="\\G", description="Where the previous match ended"),
                ],
            ),
            CheatSheetCategory(
                category=CAT_QUANTIFIERS,
                items=[
                    CheatSheetItem(character="*", description="0 or more times, greedy"),
                    CheatSheetItem(character="+", description="1 or more times, greedy"),
                    CheatSheetItem(character="?", description="0 or 1 time, greedy"),
                    CheatSheetItem(character="{m,n}", description="Between m and n times, greedy"),
                    CheatSheetItem(character="*?", description="0 or more times, lazy"),
                    CheatSheetItem(character="+?", description="1 or more times, lazy"),
                    CheatSheetItem(character="*+", description="0 or more times, possessive"),
                    CheatSheetItem(character="++", description="1 or more times, possessive"),
                    CheatSheetItem(character="?+", description="0 or 1 time, possessive"),
                ],
            ),
            CheatSheetCategory(
                category=CAT_GROUPS,
                items=[
                    CheatSheetItem(character="(...)", description="Capturing group; MySQL cannot report its position"),
                    CheatSheetItem(character="(?:...)", description="Non-capturing group"),
                    CheatSheetItem(character="x|y", description="Alternation (match x or y)"),
                    CheatSheetItem(character="\\1", description="Backreference to capture group 1"),
                    CheatSheetItem(character="$1", description="Group reference inside a REGEXP_REPLACE replacement"),
                ],
            ),
            CheatSheetCategory(
                category=CAT_ADVANCED,
                items=[
                    CheatSheetItem(character="(?=...)", description="Positive lookahead"),
                    CheatSheetItem(character="(?!...)", description="Negative lookahead"),
                    CheatSheetItem(character="(?<=...)", description="Positive lookbehind"),
                    CheatSheetItem(character="(?<!...)", description="Negative lookbehind"),
                    CheatSheetItem(character="(?>...)", description="Atomic group"),
                    CheatSheetItem(character="(?i)", description="Inline flag: case-insensitive"),
                    CheatSheetItem(character="(?s)", description="Inline flag: dot matches newline"),
                    CheatSheetItem(character="(?m)", description="Inline flag: multi-line anchors"),
                    CheatSheetItem(character="(?x)", description="Inline flag: free-spacing"),
                    CheatSheetItem(character="(?i:...)", description="Scoped inline flag group"),
                ],
            ),
            CheatSheetCategory(
                category=CAT_SQL,
                items=[
                    CheatSheetItem(character="expr REGEXP pat", description="Boolean match operator"),
                    CheatSheetItem(character="expr RLIKE pat", description="Synonym for REGEXP"),
                    CheatSheetItem(character="REGEXP_LIKE(s, p, mt)", description="Boolean match test with an explicit match type"),
                    CheatSheetItem(character="REGEXP_INSTR(s, p, pos, occ, ret, mt)", description="1-based position of a match"),
                    CheatSheetItem(character="REGEXP_SUBSTR(s, p, pos, occ, mt)", description="Text of the occ-th match"),
                    CheatSheetItem(character="REGEXP_REPLACE(s, p, r, pos, occ, mt)", description="Replace matches; $1 refers to a group"),
                    CheatSheetItem(character="SET regexp_time_limit", description="ICU step budget guarding against runaway patterns"),
                    CheatSheetItem(character="SET regexp_stack_limit", description="Memory budget for the ICU match stack"),
                ],
            ),
        ],
        engine_examples=[
            EngineExample(
                regex=r"(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])(?:\.(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])){3}(?::[0-9]{1,5})?",
                text="This is an example to get IP:\n\n192.168.1.100\n192.168.1.100:8080\n127.0.0.1\n192.168.1.0/24\n192.168.1.1-192.168.1.255",
            )
        ],
    )


def build_worker_info(server_version: str) -> WorkerInfo:
    return WorkerInfo(
        worker_name=WORKER_NAME,
        worker_version=WORKER_VERSION,
        worker_release_date=WORKER_RELEASE_DATE,
        worker_schema_version=WORKER_SCHEMA_VERSION,
        engines=[build_engine(server_version)],
    )


def register_engines(redis_client, worker_info: WorkerInfo) -> None:
    redis_client.hset("openregex:workers", worker_info.worker_name, worker_info.model_dump_json())
    print(f"[Worker] Registered '{worker_info.worker_name}' with {len(worker_info.engines)} engines.", flush=True)

File: workers/worker-ruby/src/registry.rb
# frozen_string_literal: true

require 'json'

module OpenRegex
  # Builds the engine descriptor published to the backend via Redis.
  module Registry
    WORKER_NAME = 'worker-ruby'

    FLAG_METADATA = {
      'i' => ['Case-insensitive matching.', 'Basic'],
      'm' => ['Multi-line mode. In Ruby this makes . match newline (dotall elsewhere).', 'Unique'],
      'x' => ['Extended mode. Ignores unescaped whitespace and allows comments.', 'Basic'],
      'o' => ['Interpolate the pattern only once; accepted for parity, no effect here.', 'Unique']
    }.freeze

    def self.build_flags(names)
      names.map do |name|
        description, group = FLAG_METADATA.fetch(name, ["Flag (#{name})", 'Basic'])
        { 'name' => name, 'description' => description, 'group' => group }
      end
    end

    def self.onigmo_version
      # Onigmo does not expose its version number to Ruby code; report the host runtime.
      "bundled with Ruby #{RUBY_VERSION}"
    end

    def self.worker_info
      {
        'worker_name' => WORKER_NAME,
        'worker_version' => ENV.fetch('WORKER_VERSION', 'Unknown'),
        'worker_release_date' => ENV.fetch('WORKER_RELEASE_DATE', 'Unreleased'),
        # Contract version: "1.1" = match offsets normalized to Unicode code points.
        'worker_schema_version' => '1.1',
        'engines' => [engine]
      }
    end

    def self.engine
      {
        'engine_id' => 'ruby_onigmo',
        'engine_language_type' => 'Ruby',
        'engine_language_version' => RUBY_VERSION,
        'engine_regex_lib' => 'Onigmo',
        'engine_regex_lib_version' => onigmo_version,
        'engine_label' => "Ruby #{RUBY_VERSION.split('.').first(2).join('.')} (Onigmo)",
        'engine_capabilities' => {
          'flags' => build_flags(%w[i m x o]),
          'supports_lookaround' => true,
          'supports_backrefs' => true
        },
        'engine_docs' => {
          'trivia' => [
            'Onigmo is a fork of Oniguruma maintained specifically for Ruby; it adds Perl-style constructs such as \\K and the absence operator.',
            'Onigmo is distributed under the BSD-2-Clause license; Ruby itself under the Ruby License or BSD-2-Clause.',
            "Ruby's 'm' flag is NOT multiline in the Perl sense - it is dotall. ^ and $ always match at line boundaries in Ruby.",
            'Ruby 3.2 introduced Regexp.timeout, a global ReDoS guard that aborts a match after a configurable wall-clock budget. OpenRegex runs every request under that guard.',
            'Ruby 3.2 also added a memoizing optimizer that makes many otherwise exponential patterns run in linear time.',
            'Match offsets in Ruby MatchData are character indices, never byte indices, so multibyte subjects behave intuitively.',
            'Named groups switch off numbered capturing: once (?<name>...) appears in a pattern, plain (...) groups stop capturing.'
          ],
          'cheat_sheet_url' => 'https://docs.ruby-lang.org/en/master/Regexp.html'
        },
        'engine_cheat_sheet' => cheat_sheet,
        'engine_examples' => [
          {
            'regex' => '(?<IP>(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))(?:/(?<mask>\d{1,2}))?(?::(?<port>\d{1,5}))?',
            'text' => "This is an example to get IP:\n\n192.168.1.100\n192.168.1.100:8080\n127.0.0.1\n192.168.1.0/24\n192.168.1.1-192.168.1.255"
          }
        ]
      }
    end

    def self.cheat_sheet
      [
        {
          'category' => 'Character Classes & Escapes',
          'items' => [
            { 'character' => '.', 'description' => "Any character except newline unless the 'm' flag is set" },
            { 'character' => '\\w', 'description' => 'Word character; ASCII-only unless the pattern is Unicode-encoded' },
            { 'character' => '\\W', 'description' => 'Non-word character' },
            { 'character' => '\\d', 'description' => 'Decimal digit (ASCII)' },
            { 'character' => '\\D', 'description' => 'Non-digit' },
            { 'character' => '\\s', 'description' => 'Whitespace character' },
            { 'character' => '\\S', 'description' => 'Non-whitespace character' },
            { 'character' => '\\h', 'description' => 'Hexadecimal digit character' },
            { 'character' => '\\R', 'description' => 'Generic linebreak (\\r\\n, \\n, \\v, \\f, \\r, U+0085, U+2028, U+2029)' },
            { 'character' => '\\p{Alpha}', 'description' => 'POSIX or Unicode property class' },
            { 'character' => '\\p{Han}', 'description' => 'Unicode script property' },
            { 'character' => '[[:alpha:]]', 'description' => 'POSIX bracket expression' },
            { 'character' => '[a-z&&[^aeiou]]', 'description' => 'Character class intersection' }
          ]
        },
        {
          'category' => 'Anchors & Boundaries',
          'items' => [
            { 'character' => '^', 'description' => 'Start of line - always line-based in Ruby' },
            { 'character' => '$', 'description' => 'End of line - always line-based in Ruby' },
            { 'character' => '\\A', 'description' => 'Start of string' },
            { 'character' => '\\z', 'description' => 'End of string' },
            { 'character' => '\\Z', 'description' => 'End of string, ignoring a trailing newline' },
            { 'character' => '\\b', 'description' => 'Word boundary' },
            { 'character' => '\\B', 'description' => 'Non-word boundary' },
            { 'character' => '\\G', 'description' => 'Where the previous match ended' },
            { 'character' => '\\K', 'description' => 'Keep-out: drop everything matched so far from the reported match' }
          ]
        },
        {
          'category' => 'Quantifiers',
          'items' => [
            { 'character' => '*', 'description' => '0 or more times, greedy' },
            { 'character' => '+', 'description' => '1 or more times, greedy' },
            { 'character' => '?', 'description' => '0 or 1 time, greedy' },
            { 'character' => '{m,n}', 'description' => 'Between m and n times, greedy' },
            { 'character' => '*?', 'description' => '0 or more times, lazy' },
            { 'character' => '+?', 'description' => '1 or more times, lazy' },
            { 'character' => '??', 'description' => '0 or 1 time, lazy' },
            { 'character' => '*+', 'description' => '0 or more times, possessive' },
            { 'character' => '++', 'description' => '1 or more times, possessive' },
            { 'character' => '?+', 'description' => '0 or 1 time, possessive' }
          ]
        },
        {
          'category' => 'Grouping & Backreferences',
          'items' => [
            { 'character' => '(...)', 'description' => 'Capturing group' },
            { 'character' => '(?:...)', 'description' => 'Non-capturing group' },
            { 'character' => 'x|y', 'description' => 'Alternation (match x or y)' },
            { 'character' => '(?<name>...)', 'description' => 'Named capturing group; disables numbered capture' },
            { 'character' => "(?'name'...)", 'description' => 'Named capturing group, quoted syntax' },
            { 'character' => '\\1', 'description' => 'Backreference to capture group 1' },
            { 'character' => '\\k<name>', 'description' => 'Backreference to a named group' },
            { 'character' => '\\k<-1>', 'description' => 'Relative backreference to the previous group' },
            { 'character' => '\\g<name>', 'description' => 'Subexpression call (recursion into a named group)' },
            { 'character' => '\\g<0>', 'description' => 'Recursive call of the whole pattern' }
          ]
        },
        {
          'category' => 'Lookarounds & Advanced',
          'items' => [
            { 'character' => '(?=...)', 'description' => 'Positive lookahead' },
            { 'character' => '(?!...)', 'description' => 'Negative lookahead' },
            { 'character' => '(?<=...)', 'description' => 'Positive lookbehind' },
            { 'character' => '(?<!...)', 'description' => 'Negative lookbehind' },
            { 'character' => '(?>...)', 'description' => 'Atomic group; prevents backtracking' },
            { 'character' => '(?~...)', 'description' => 'Absence operator: match anything that does not contain the subpattern' },
            { 'character' => '(?(1)yes|no)', 'description' => 'Conditional on whether group 1 participated' },
            { 'character' => '(?i)', 'description' => 'Inline flag: case-insensitive' },
            { 'character' => '(?m)', 'description' => 'Inline flag: dot matches newline' },
            { 'character' => '(?x)', 'description' => 'Inline flag: extended / free-spacing' },
            { 'character' => '(?i:...)', 'description' => 'Scoped inline flag group' },
            { 'character' => '(?-i:...)', 'description' => 'Scoped inline flag group, flag turned off' }
          ]
        }
      ]
    end
  end
end
